fbpic/boundaries/field_buffer_handling.py

```diff
@@ -31,301 +34,449 @@ def __init__( self, n_guard, Nr, Nm, left_proc, right_proc ):
 
         Depending on whether the field data is initially on the CPU
         or on the GPU, this function will do the appropriate exchange
         with the device.
 
         Parameters
         ----------
-        interp: a list of InterpolationGrid objects
-            (one element per azimuthal mode)
+        grid_m_x: InterpolationGrid objects
+            6 Interpolation grid objects. One for each of the two modes (0, 1)
+            and for each coordinate (r, t, z). (m = mode, x = coordinate)
+
+        method: str
+            Can either be 'replace' or 'add' depending on the type
+            of field exchange that is needed
+
+        use_cuda: bool
+            Whether the simulation runs on GPUs. If True,
+            the buffers are copied to the GPU arrays after the MPI exchange.
 
         before_sending: bool
             Whether to copy the inner part of the domain to the sending buffer
 
         after_receiving: bool
             Whether to copy the receiving buffer to the guard cells
         """
-        # Shortcut for the guard cells
-        ng = self.n_guard
+        # Define region that is copied to or from the buffer
+        # depending on the method used.
+        if method == 'replace':
+            nz_start = self.n_guard
+            nz_end = 2*self.n_guard
+        if method == 'add':
+            nz_start = 0
+            nz_end = 2*self.n_guard
+        # Whether or not to send to the left or right neighbor
         copy_left = (self.left_proc is not None)
         copy_right = (self.right_proc is not None)
 
         # When using the GPU
-        if interp[0].use_cuda:
-
+        if use_cuda:
             # Calculate the number of blocks and threads per block
-            dim_grid_2d, dim_block_2d = cuda_tpb_bpg_2d( ng, interp[0].Nr )
+            dim_grid_2d, dim_block_2d = cuda_tpb_bpg_2d(
+                nz_end - nz_start, self.Nr )
 
             if before_sending:
-                # Copy the inner regions of the domain to the GPU buffers
-                copy_EB_to_gpu_buffers[ dim_grid_2d, dim_block_2d ](
-                    self.d_EB_l, self.d_EB_r,
-                    interp[0].Er, interp[0].Et, interp[0].Ez,
-                    interp[0].Br, interp[0].Bt, interp[0].Bz,
-                    interp[1].Er, interp[1].Et, interp[1].Ez,
-                    interp[1].Br, interp[1].Bt, interp[1].Bz,
-                    copy_left, copy_right, ng )
-                # Copy the GPU buffers to the sending CPU buffers
-                if copy_left:
-                    self.d_EB_l.copy_to_host( self.EB_send_l )
-                if copy_right:
-                    self.d_EB_r.copy_to_host( self.EB_send_r )
+                if method == 'replace':
+                    # Copy the inner regions of the domain to the buffers
+                    copy_vec_to_gpu_buffer[ dim_grid_2d, dim_block_2d ](
+                        self.d_vec_rep_buffer_l, self.d_vec_rep_buffer_r,
```

Comment:
It seems that the variables `self.d_vec_rep_buffer_l` and `self.d_vec_rep_buffer_r` do not exist and cause the code to crash, when executed on GPU. 

@MKirchen : Could you try to run a test on GPU with MPI, and confirm that you also get an error?